# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/CodeGenerator.cpp
# Commit: a0e51f584680
# Full Hash: a0e51f584680a017fc4d8124c58a51f3caadfade
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2021-09-16 09:43:03
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Introduce indirect stubs to optimize call_indirect. r=lth
#   
#   This patch introduces indirect stubs.
#   An indirect stub is a stub that takes care of any switching activities needed for call_indirect.
#   Before this patch, we have to conservatively assume that any call_indirect's target can be from a foreign instance,
# ==============================================================================

diff -r ece3dcc343a2 -r a0e51f584680 js/src/jit/CodeGenerator.cpp
--- a/js/src/jit/CodeGenerator.cpp	Thu Sep 16 04:38:26 2021 +0000
+++ b/js/src/jit/CodeGenerator.cpp	Thu Sep 16 05:09:24 2021 +0000
@@ -7920,37 +7920,34 @@
 #endif
 
   // LWasmCallBase::isCallPreserved() assumes that all MWasmCalls preserve the
-  // TLS and pinned regs. The only case where where we don't have to reload
-  // the TLS and pinned regs is when the callee preserves them.
-  bool reloadRegs = true;
-  bool switchRealm = true;
+  // TLS and pinned regs. The only case where where we have to reload
+  // the TLS and pinned regs is when we emit import or builtin instance calls.
+  bool reloadRegs = false;
+  bool switchRealm = false;
 
   const wasm::CallSiteDesc& desc = mir->desc();
   const wasm::CalleeDesc& callee = mir->callee();
   switch (callee.which()) {
     case wasm::CalleeDesc::Func:
       masm.call(desc, callee.funcIndex());
-      reloadRegs = false;
-      switchRealm = false;
       break;
     case wasm::CalleeDesc::Import:
       masm.wasmCallImport(desc, callee);
+      reloadRegs = true;
+      switchRealm = true;
       break;
     case wasm::CalleeDesc::AsmJSTable:
     case wasm::CalleeDesc::WasmTable:
       masm.wasmCallIndirect(desc, callee, needsBoundsCheck);
-      reloadRegs = switchRealm = callee.which() == wasm::CalleeDesc::WasmTable;
       break;
     case wasm::CalleeDesc::Builtin:
       masm.call(desc, callee.builtin());
-      reloadRegs = false;
-      switchRealm = false;
       break;
     case wasm::CalleeDesc::BuiltinInstanceMethod:
       masm.wasmCallBuiltinInstanceMethod(desc, mir->instanceArg(),
                                          callee.builtin(),
                                          mir->builtinMethodFailureMode());
-      switchRealm = false;
+      reloadRegs = true;
       break;
   }
 